src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -231,51 +259,232 @@ def frobenius_charpoly(self, var='X'):
+            sage: phi._frobenius_charpoly_crystalline('X')
+            X^11 + (z3^2 + 2*z3)*X^10 + ((z3 + 1)*T + z3)*X^9 + ((2*z3^2 + z3 + 2)*T^2 + ... + (2*z3^2 + 2*z3 + 2)*T + z3^2
 
         ALGORITHM:
 
-            We compute the Frobenius norm, and with it the Frobenius
-            trace. This gives the Frobenius characteristic polynomial.
-            See [MS2019]_, Section 4.
+        Compute the characteristic polynomial of the Frobenius endomorphism
+        acting on the crystalline cohomology of a Drinfeld module, which
+        is equal to that of the Frobenius endomorphism on the Drinfeld
+        module. A recurrence on elements of the cohomology allows us to
+        compute a matrix representation of the Frobenius endomorphism
+        efficiently using a companion matrix method.
+        """
+        Fq = self._Fq
+        K = self.base_over_constants_field()
+        A = self.function_ring()
+        char, q = Fq.characteristic(), Fq.cardinality()
+        qdeg = logb(q, char)
+        r, n = self.rank(), self._base_degree_over_constants
+        nstar = ceil(sqrt(n))
+        nquo, nrem = divmod(n, nstar)
+        drin_coeffs = self.coefficients(sparse=False)
+        poly_K = PolynomialRing(K, name=str(A.gen()))
+        matrix_poly_K = MatrixSpace(poly_K, r, r)
+        mu_coeffs = ((poly_K.gen() - drin_coeffs[0])**(n+1)).coefficients(
+                                                             sparse=False)
+
+        def companion(order):
+            # + [1] is required to satisfy formatting for companion matrix
+            M = matrix_poly_K(companion_matrix([(drin_coeffs[i]/drin_coeffs[r])
+                               .frobenius(qdeg*order)
+                               for i in range(r)] + [1], format='top'))
+            M[0, r-1] += poly_K.gen() / drin_coeffs[r].frobenius(qdeg*order)
+            return M
+
+        companion_initial = prod([companion(i) for i in range(nrem, 0, -1)])
+        companion_step = prod([companion(i)
+                              for i in range(nstar + nrem, nrem, -1)])
+        reduced_companions = []
+        for k in range(nquo - 1, 0, -1):
+            M = Matrix(poly_K, r, r)
+            for i, row in enumerate(companion_step):
+                for j, entry in enumerate(row):
+                    reduction = entry % poly_K([c.frobenius(
+                                                qdeg*(-k*nstar % n))
+                                                for c in mu_coeffs])
+                    M[i, j] = poly_K([c.frobenius(qdeg*(k*nstar))
+                                     for c in reduction.coefficients(
+                                              sparse=False)])
+            reduced_companions.append(M)
+        charpoly_coeffs_K = (prod(reduced_companions) * companion_step *
+                             companion_initial).charpoly(var).coefficients(
+                             sparse=False)
```

Comment:
More precisely, my point is the following: I agree that cutting lines could be nice but, in this case, I think that the locations when the cut are done are not meaningful (and even confusing).

Reply:
Yeah OK. This example is terrible, and you are definitely right on this one. However putting `.coefficients(sparse=False)` on a single line would exceed the 79 chars limit. It's not the kind of things I like to break. If you change it sneakily, I will not notice.

Reply:
Let me also recall that creating intermediate variables (with meaningful names!) could also be a good option to break long lines. (And it's not less performant.)

Reply:
Here for instance, it could be
```sage
    mat = prod(reduced_companions) * companion_step * companion_initial
    charpoly_coeffs_K = mat.charpoly(var).list()
```
(which is even 2 lines instead of 3).

Reply:
It is *way* better. Thank you.